### DISCUSSION TITLE
Logging of next retry in RetryTopicConfiguration - Spring kafka retry

### DISCUSSION THREAD

**Discussion by hariraogotit** [spring-projects/spring-kafka#2992] [Q&A] [answered]
Team,
  I am using spring kafka retry - RetryTopicConfiguration. A question, any lead is much appreciated please.
  
  Is there a way to extract when will be the next retry?
  
  eg say we have   2 retry topics topic-2000, topic-6000
  
  We know that when it writes to topic-2000, it will process after 2000ms and similar when it writes to 6000, it will process after 6000ms but is there a way to extract information about how long it will take for retry to kick in?
  
  I want to log something like retry will happen after 2000ms. If it is not successful then I want to log that retry will happen in 6000ms.

**Comment by hariraogotit** [comment]
> Team, I am using spring kafka retry - RetryTopicConfiguration. A question, any lead is much appreciated please.
> 
> Is there a way to extract when will be the next retry?
> 
> eg say we have 2 retry topics topic-2000, topic-6000
> 
> We know that when it writes to topic-2000, it will process after 2000ms and similar when it writes to 6000, it will process after 6000ms but is there a way to extract information about how long it will take for retry to kick in?
> 
> I want to log something like retry will happen after 2000ms. If it is not successful then I want to log that retry will happen in 6000ms.


  
  i see the line "if a message processing throws an exception, the configured [DefaultErrorHandler](https://docs.spring.io/spring-kafka/api/org/springframework/kafka/listener/DefaultErrorHandler.html) and [DeadLetterPublishingRecoverer](https://docs.spring.io/spring-kafka/api/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.html) forwards the message to the next topic, using a [DestinationTopicResolver](https://docs.spring.io/spring-kafka/api/org/springframework/kafka/retrytopic/DestinationTopicResolver.html) to know the next topic and the delay for it."
  
  in https://docs.spring.io/spring-kafka/api/org/springframework/kafka/retrytopic/RetryTopicConfigurer.html
  
  And I see List<DestinationTopic.Properties> destinationTopicProperties; in RetryTopicConfiguration
  
  But when I follow along I am not able to get hold of DestinationTopicResolver.

**Comment by artembilan** [comment]
The `KafkaBackoffAwareMessageListenerAdapter` has a logic like this:
```
		maybeGetBackoffTimestamp(consumerRecord)
				.ifPresent(nextExecutionTimestamp -> this.kafkaConsumerBackoffManager
						.backOffIfNecessary(createContext(consumerRecord, nextExecutionTimestamp, consumer)));
```
and that one does this:
```
	public void backOffIfNecessary(Context context) {
		long backoffTime = context.getDueTimestamp() - System.currentTimeMillis();
		LOGGER.debug(() -> "Back off time: " + backoffTime + " Context: " + context);
		if (backoffTime > 0) {
			pauseConsumptionAndThrow(context, backoffTime);
		}
	}
```
Isn't that debug message not enough for you?

**Reply by artembilan** [reply]
You don't. It is done by the `ListenerContainerFactoryConfigurer` , which is used from the `RetryTopicComponentFactory`.
And that one is used by the `RetryTopicBeanNames.RETRY_TOPIC_CONFIGURER_BEAN_NAME` bean from  the `RetryTopicConfigurationSupport`.

**Reply by hariraogotit** [reply]
@artembilan got you but in PROD or in the higher envs' we have INFO level logging. So this will not be logged.

**Reply by artembilan** [reply]
Well, in prod I prefer to have it as a `warn`. If application works, then no need to log anything.
Why would one look into prod logs for some info if it does not fail?

**Reply by hariraogotit** [reply]
Basically I want to log after how many more ms it is going to retry. This will be  helpful in terms of trouble shooting. 

https://docs.spring.io/spring-kafka/reference/retrytopic/access-topic-info-runtime.html

talks about DestinationTopicResolver. Will this help? If so how I do I get to this from RetryTopicConfigurationBuilder?

Thanks for taking time @artembilan .

**Reply by artembilan** [reply]
Why do you need to get access to that from the `RetryTopicConfigurationBuilder`?
This `DestinationTopicResolver` is created as a bean from the `RetryTopicConfigurationSupport`.
So, you get access to that bean from whenever you need and do whatever that doc describes.

Looks like that `DestinationTopicContainer` has an API:
```
	/**
	 * Returns the {@link DestinationTopic} instance registered as the next
	 * destination topic in the chain for the given topic.
	 * Note that this might not correspond to the actual next topic a message will
	 * be forwarded to, since that depends on different factors.
	 *
	 * If you need to find out the exact next topic for a message use the
	 * {@link DestinationTopicResolver#resolveDestinationTopic(String, String, Integer, Exception, long)}
	 * method instead.
	 * @param mainListenerId the listener id.
	 * @param topicName the topic name of the DestinationTopic to be returned.
	 * @return the next DestinationTopic in the chain registered for that topic.
	 */
	DestinationTopic getNextDestinationTopicFor(String mainListenerId, String topicName);
```

This `DestinationTopic` has a property:
```
	public Long getDestinationDelay() {
```

Does that help somehow?

Sorry, I'm not good with this Retry Topic subject to give any advice for customization.

**Reply by hariraogotit** [reply]
@artembilan as per the document https://docs.spring.io/spring-kafka/reference/retrytopic/retry-topic-lcf.html

RetryTopicConfiguration must be used to configure the non blocking retry and that is what I did using RetryTopicConfigurationBuilder.

**Reply by artembilan** [reply]
Correct. But how is that related to the logging concern you are talking about?
If you want to customize it somehow, then I don't see anything where we could hook at the moment of execution, unless you can deal with the debug message I pointed before.

**Reply by hariraogotit** [reply]
If RetryTopicConfiguration is the way to configure then how to stand up RetryTopicConfigurationSupport ?  I mean it looks like RetryTopicConfigurationSupport needs to be configured instead of RetryTopicConfiguration and are there any document please?

**Reply by artembilan** [reply]
It is there by `@EnableRetryTopic`.
Why do you need  to override it?
Well, you indeed can do that. See docs: https://docs.spring.io/spring-kafka/reference/retrytopic/retry-config.html#retry-topic-global-settings
And there you can override any of its factory methods to whatever you'd like. Including that `DestinationTopicResolver`.

**Reply by hariraogotit** [reply]
I am trying to figure out a way to get hold of the object DestinationTopicResolver.  Will read the documentation you pointed out. Thanks a lot for your time so far.

**Reply by hariraogotit** [reply]
@artembilan in DeadLetterPublishingRecovererFactory I see the following code 

Are these metadata reliable and sort my issue as I can simply extract info from these headers (DEFAULT_HEADER_BACKOFF_TIMESTAMP, DEFAULT_HEADER_ATTEMPTS and DEFAULT_HEADER_ORIGINAL_TIMESTAMP) and log?

```
private Headers addHeaders(String mainListenerId, ConsumerRecord<?, ?> consumerRecord, Exception e, int attempts) {
		Headers headers = new RecordHeaders();
		byte[] originalTimestampHeader = getOriginalTimestampHeaderBytes(consumerRecord);
		if (!this.retainAllRetryHeaderValues) {
			headers.add(new SingleRecordHeader(RetryTopicHeaders.DEFAULT_HEADER_ORIGINAL_TIMESTAMP,
					originalTimestampHeader));
			headers.add(new SingleRecordHeader(RetryTopicHeaders.DEFAULT_HEADER_ATTEMPTS,
					ByteBuffer.wrap(new byte[Integer.BYTES]).putInt(attempts + 1).array()));
			headers.add(new SingleRecordHeader(RetryTopicHeaders.DEFAULT_HEADER_BACKOFF_TIMESTAMP,
					BigInteger
							.valueOf(getNextExecutionTimestamp(mainListenerId, consumerRecord, e,
									originalTimestampHeader))
							.toByteArray()));
		}
		else {
			headers.add(RetryTopicHeaders.DEFAULT_HEADER_ORIGINAL_TIMESTAMP, originalTimestampHeader);
			headers.add(RetryTopicHeaders.DEFAULT_HEADER_ATTEMPTS,
					ByteBuffer.wrap(new byte[Integer.BYTES]).putInt(attempts + 1).array());
			headers.add(RetryTopicHeaders.DEFAULT_HEADER_BACKOFF_TIMESTAMP,
					BigInteger
							.valueOf(getNextExecutionTimestamp(mainListenerId, consumerRecord, e,
									originalTimestampHeader))
							.toByteArray());
		}
		return headers;
	}
```

**Reply by artembilan** [reply]
So, that `DeadLetterPublishingRecovererFactory` populates those headers which then used in the `ContainerPartitionPausingBackOffManager` to pause the partition via `BackOffHandler` for the time provided in the `DEFAULT_HEADER_BACKOFF_TIMESTAMP`.

If you see how you can use that info for yourself, then yes, you can try to rely on it.

**Reply by hariraogotit** [reply]
In all of my KafkaListeners I am planning to extract from the headers as below

Header ot = consumerRecord.headers().lastHeader(RetryTopicHeaders.DEFAULT_HEADER_ORIGINAL_TIMESTAMP);
Header count = consumerRecord.headers().lastHeader(RetryTopicHeaders.DEFAULT_HEADER_ATTEMPTS);
Header bt = consumerRecord.headers().lastHeader(RetryTopicHeaders.DEFAULT_HEADER_BACKOFF_TIMESTAMP);

**Reply by artembilan** [reply]
That's no OK. The record will reach your listener only after back off. 
So, even if you receive the info eventually, it does not give you a desired effect you have talked in the beginning for determining when the next retry after delay.

**Reply by hariraogotit** [reply]
I tried this in a play ground project  with one kafka listener.

When I send a proper message the headers are all null but when I send a poison pill (bad data - hari sydney). The headers ot, count and bt are not null.

Count gives the count, I am able to get the original time and back off time i,e it looks like it is the time when the next retry will kick in. Some logs below 

```
Topic name from listener Sydney -> APZ-sydney
Countnull
okay
Message from the listener Sydney --> hari sydney

Topic name from listener Sydney -> APZ-sydney-AP-Australia-retry-1000
Original time stamp 1705597193130
2024-01-18T16:59:53.130
CountRecordHeader(key = retry_topic-attempts, value = [0, 0, 0, 2])
Back off time 1705597195153
2024-01-18T16:59:55.153
okay
Message from the listener Sydney --> hari sydney

2024-01-18T16:59:55.697Z  INFO 70197 --- [ad | producer-1] org.apache.kafka.clients.Metadata        : [Producer clientId=producer-1] Resetting the last seen epoch of partition APZ-sydney-AP-Australia-retry-2000-0 to 0 since the associated topicId changed from null to DSL0fzmMRouv97MUEAd6Nw
2024-01-18T16:59:55.703Z  INFO 70197 --- [etry-2000-0-C-1] o.a.k.clients.consumer.KafkaConsumer     : [Consumer clientId=consumer-sydney-AP-Australia-retry-2000-2, groupId=sydney-AP-Australia-retry-2000] Seeking to offset 38 for partition APZ-sydney-AP-Australia-retry-2000-0
Topic name from listener Sydney -> APZ-sydney-AP-Australia-retry-2000
Original time stamp 1705597193130
2024-01-18T16:59:53.130
CountRecordHeader(key = retry_topic-attempts, value = [0, 0, 0, 3])
Back off time 1705597197196
2024-01-18T16:59:57.196
okay
Message from the listener Sydney --> hari sydney

2024-01-18T16:59:57.709Z  INFO 70197 --- [ad | producer-1] org.apache.kafka.clients.Metadata        : [Producer clientId=producer-1] Resetting the last seen epoch of partition APZ-sydney-AP-Australia-retry-4000-0 to 0 since the associated topicId changed from null to YvYG68RZSjC-YSddFX4lOg
2024-01-18T16:59:57.713Z  INFO 70197 --- [etry-4000-0-C-1] o.a.k.clients.consumer.KafkaConsumer     : [Consumer clientId=consumer-sydney-AP-Australia-retry-4000-4, groupId=sydney-AP-Australia-retry-4000] Seeking to offset 39 for partition APZ-sydney-AP-Australia-retry-4000-0
Topic name from listener Sydney -> APZ-sydney-AP-Australia-retry-4000
Original time stamp 1705597193130
2024-01-18T16:59:53.130
CountRecordHeader(key = retry_topic-attempts, value = [0, 0, 0, 4])
Back off time 1705597201202
2024-01-18T17:00:01.202
okay
Message from the listener Sydney --> hari sydney


2024-01-18T17:00:01.209Z ERROR 70197 --- [etry-4000-0-C-1] k.r.DeadLetterPublishingRecovererFactory : Record: topic = APZ-sydney-AP-Australia-retry-4000, partition = 0, offset = 39, main topic = APZ-sydney threw an error at topic APZ-sydney-AP-Australia-retry-4000 and won't be retried. Sending to DLT with name APZ-sydney-AP-Australia-dlt.

```

I converted the long timestamp to a readable one. 
Eg the back off 2024-01-18T16:59:55.153 is the next retry time as you can see the next retry is attempted as shown below on the specified time
`
2024-01-18T16:59:55.697Z  INFO 70197 --- [ad | producer-1] org.apache.kafka.clients.Metadata `

**Reply by artembilan** [reply, marked as answer]
OK. It is still hard for me to understand what is going on, but sounds like we can close this discussion since you got the answer for your original request.

**Reply by hariraogotit** [reply]
Apologies I formatted it now

basically the below i.e back off time gives when the next retry will kick in and next retry happens at that time 

```
Back off time 1705597197196
2024-01-18T16:59:57.196
```


`2024-01-18T16:59:57.709Z  INFO 70197 --- [ad | producer-1] org.apache.kafka.clients.Metadat`
